File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-6.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        res = False 

        def dfs(node):  
            nonlocal res
            if not node: 
                return None 

            if node.val == subRoot.val: 
                if (self.isSameTree(node, subRoot)): 
                    res = True 

            dfs(node.left) 
            dfs(node.right) 

        dfs(root) 
        return res

    def isSameTree(self, n1, n2): 
        if not n1 and not n2: 
            return True 
        elif n1 and n2 and n1.val == n2.val:
            return (self.isSameTree(n1.left, n2.left) and self.isSameTree(n1.right, n2.right)) 
        else: 
            return False

File: Data Structures & Algorithms/same-binary-tree/submission-6.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {boolean}
     */
     isSameTree(p, q) { 

        const stack = [[p, q]] 
         

         while(stack.length) {
            const [n1, n2] = stack.pop() 

            if (!n1 && !n2) continue 

            if (!n1 || !n2 || n1.val !== n2.val) return false 

            stack.push([n1.left, n2.left]) 
            stack.push([n1.right, n2.right])
         } 

         return true
    } 
}
